tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java

```diff
@@ -0,0 +1,222 @@
+package tlc2.debug;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import tla2sany.parser.SyntaxTreeNode;
+import tla2sany.parser.TLAplusParser;
+import tla2sany.semantic.AbortException;
+import tla2sany.semantic.Context;
+import tla2sany.semantic.Errors;
+import tla2sany.semantic.FormalParamNode;
+import tla2sany.semantic.Generator;
+import tla2sany.semantic.LetInNode;
+import tla2sany.semantic.LevelConstants;
+import tla2sany.semantic.ModuleNode;
+import tla2sany.semantic.OpApplNode;
+import tla2sany.semantic.OpDefNode;
+import tla2sany.semantic.SemanticNode;
+import tla2sany.st.Location;
+import tlc2.tool.impl.SpecProcessor;
+import util.ToolIO;
+
+/**
+ * An expression for conditionally triggering a breakpoint.
+ */
+public class TLCBreakpointExpression {
+
+	/**
+	 * Given a spec and an unparsed expression, build an operator that can be
+	 * evaluated within the context of that spec.
+	 *
+	 * @param processor     Processes expression within the model context.
+	 * @param semanticRoot  The root of the spec's semantic parse tree.
+	 * @param location      The breakpoint location.
+	 * @param conditionExpr The unparsed breakpoint expression.
+	 * @return A breakpoint expression, or null if parsing failed.
+	 */
+	public static OpDefNode process(
+			final SpecProcessor processor,
+			final ModuleNode semanticRoot,
+			final Location location,
+			final String conditionExpr
+	) {
+		if (null == conditionExpr || conditionExpr.isBlank()) {
+			return null;
+		}
+		ToolIO.out.println("BPExpr: processing expression \"" + conditionExpr + "\"");
+
+		final String rootModName = semanticRoot.getName().toString();
+		final String bpModName = generateUnusedName(semanticRoot, "__BreakpointModule__%s");
+		ToolIO.out.println("BPExpr: wrapping with module \"" + bpModName + "\"");
+		final String bpOpName = generateUnusedName(semanticRoot, "__BreakpointExpr__%s");
+		final List<String> paramNames = getScopedIdentifiers(semanticRoot, location);
+		final String params = paramNames.size() > 0 ? "(" + String.join(", ", paramNames) + ")" : "";
+		final String bpOpDef = bpOpName + params;
+		ToolIO.out.println("BPExpr: wrapping with op \"" + bpOpDef + "\"");
+
+		final String wrapper = "---- MODULE %s ----\nEXTENDS %s\n%s == %s\n====";
+		String wrappedConditionExpr = String.format(wrapper, bpModName, rootModName, bpOpDef, conditionExpr);
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
```

Comment:
Should this stream be closed explicitly? Perhaps, be wrapped in a “try-with-resources” block?

Reply:
Per Java docs closing a `ByteArrayInputStream` has no effect: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/ByteArrayInputStream.html#close()